File: Namrata-Sanger/307. Range Sum Query - Mutable.js
/*
Day 15
307. Range Sum Query - Mutable
Link: https://leetcode.com/problems/range-sum-query-mutable/
Level: Medium
*/

var NumArray = function (nums) {
    this.arr = nums;
    this.sum = 0;

    for (let val of this.arr) { this.sum += val }
};

NumArray.prototype.update = function (index, val) {
    this.sum = this.sum - this.arr[index] + val;
    this.arr[index] = val;
    return null;
};

NumArray.prototype.sumRange = function (left, right) {
    let totalElementRequired = right - left + 1;

    if (this.arr.length == totalElementRequired) { return this.sum }
    else {
        let s1 = 0;
        let s2 = 0;
        let total = 0;

        for (let i = 0; i < left; i++) {
            s1 += this.arr[i];
        }
        for (let j = (right + 1); j < this.arr.length; j++) {
            s2 += this.arr[j];
        }

        total = this.sum - (s1 + s2);
        return total
    }
};

File: Lokesh_Saana/307-range-sum-query-mutable.py
# https://leetcode.com/problems/range-sum-query-mutable/

import numpy as np


class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = np.array([0]+nums)
        self.x = np.zeros(len(nums)+1, dtype=int)
        for i in range(1, len(self.nums)):
            self.x[i] = self.nums[i]+self.x[i-1]

    def update(self, index: int, val: int) -> None:
        self.x[index+1:] -= (self.nums[index+1] - val)
        self.nums[index+1] = val

    def sumRange(self, left: int, right: int) -> int:
        return self.x[right+1] - self.x[left]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)
